Lombiq.HelpfulLibraries/Libraries/Mvc/TypedRoute.cs

```diff
@@ -86,17 +95,20 @@ private string GetRoute(string route)
 
         public static TypedRoute CreateFromExpression<TController>(
             Expression<Action<TController>> actionExpression,
-            IEnumerable<(string Key, object Value)> additionalArguments,
+            IList<(string Key, object Value)> additionalArguments,
             ITypeFeatureProvider typeFeatureProvider = null) =>
             CreateFromExpression(
                 actionExpression,
-                additionalArguments
-                    .Select((key, value) => new KeyValuePair<string, string>(key, value.ToString())),
+                additionalArguments.Any()
+                    ? additionalArguments
+                        .Select((key, value) => new KeyValuePair<string, string>(key, value.ToString()))
+                        .ToList()
+                    : Array.Empty<KeyValuePair<string, string>>(),

```

Comment:
I wanted to avoid the cost of an unnecessary object creation. Though calling `ToList` on an empty collection is the same as a `new List<T>()` which is just a class containing an empty array anyway. Perhaps this falls into the category of  "useless micro-optimization".

> I was stunned that you could return a List from one branch, and an Array from the other, and the compiler would not complain - seems to be a C# 8 or higher language feature.

Yeah, I think type inference has been improved in C# 8. It works now becuase the argument is a known common interface so that's what the conditional arms are implicitly cast . Even before that this would've worked if you explicitly cast either arm of the conditional expression into `IList<T>`.

Reply:
> I wanted to avoid the cost of an unnecessary object creation. Though calling `ToList` on an empty collection is the same as a `new List<T>()` which is just a class containing an empty array anyway. Perhaps this falls into the category of "useless micro-optimization".

Well, it's not useless. I see this more in the category "simple code" vs. "minor (micro) performance optimization". I'm generally eager to omit object initialization when I can, so this is a good example.

OK. Let's keep your code - it is very readable, slightly more performant, and this is not about writing the least code possible.

Reply:
Now that I've converted the argument back to Enumerable that would yield multiple enumeration so I have to go with your suggestion.

Reply:
OK, makes sense.